components/feature/session/src/main/java/mozilla/components/feature/session/prompt/ChoiceDialogFragment.kt

```diff
@@ -0,0 +1,135 @@
+import android.support.v7.widget.LinearLayoutManager
+import android.support.v7.widget.RecyclerView
+import android.view.LayoutInflater
+import android.view.View
+import mozilla.components.concept.engine.prompt.Choice
+import mozilla.components.feature.session.R
+
+private const val KEY_CHOICES = "KEY_CHOICES"
+private const val KEY_DIALOG_TYPE = "KEY_DIALOG_TYPE"
+
+/**
+ * [android.support.v4.app.DialogFragment] implementation to display choice(options,optgroup and menu)
+ * web content in native dialogs.
+ */
+class ChoiceDialogFragment : PromptDialogFragment() {
+
+    @VisibleForTesting
+    internal val choices: Array<Choice> by lazy { safeArguments.getParcelableArray(KEY_CHOICES)!!.toArrayOfChoices() }
+
+    @VisibleForTesting
+    internal val dialogType: Int by lazy { safeArguments.getInt(KEY_DIALOG_TYPE) }
+
+    internal val isSingleChoice get() = dialogType == SINGLE_CHOICE_DIALOG_TYPE
+
+    internal val isMenuChoice get() = dialogType == MENU_CHOICE_DIALOG_TYPE
+
+    internal val mapSelectChoice by lazy { HashMap<Choice, Choice>() }
+
+    private val safeArguments get() = requireNotNull(arguments)
+
+    @VisibleForTesting
+    internal var isATest = false
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        return when (dialogType) {
+            SINGLE_CHOICE_DIALOG_TYPE -> createSingleChoiceDialog()
+            MULTIPLE_CHOICE_DIALOG_TYPE -> createMultipleChoiceDialog()
+            MENU_CHOICE_DIALOG_TYPE -> createSingleChoiceDialog()
+            else -> throw IllegalArgumentException(" $dialogType is not a valid choice dialog type")
+        }
+    }
+
+    companion object {
+
+        fun newInstance(choices: Array<Choice>, sessionId: String, dialogType: Int): ChoiceDialogFragment {
+            val fragment = ChoiceDialogFragment()
+            val arguments = fragment.arguments ?: Bundle()
+
+            with(arguments) {
+                putParcelableArray(KEY_CHOICES, choices)
+                putString(KEY_SESSION_ID, sessionId)
+                putInt(KEY_DIALOG_TYPE, dialogType)
+            }
+
+            fragment.arguments = arguments
+
+            return fragment
+        }
+
+        const val SINGLE_CHOICE_DIALOG_TYPE = 0
+        const val MULTIPLE_CHOICE_DIALOG_TYPE = 1
+        const val MENU_CHOICE_DIALOG_TYPE = 2
+    }
+
+    @SuppressLint("InflateParams")
+    internal fun createDialogContentView(inflater: LayoutInflater): View {
+        val view = inflater.inflate(R.layout.mozac_feature_choice_dialogs, null)
+        view.findViewById<RecyclerView>(R.id.recyclerView).apply {
+            layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
+            adapter = ChoiceAdapter(this@ChoiceDialogFragment, inflater)
+        }
+        return view
+    }
+
+    fun onSelect(selectedChoice: Choice) {
+        feature?.onSingleChoiceSelect(sessionId, selectedChoice)
+
+        if (!isATest) {
+            dismiss()
+        }
```

Comment:
This is dangerous since you start testing something else than what is running in production. If you do not want to call `dismiss()` in tests, can you stub it? E.g.:

```Kotlin
val fragment = spy(ChoiceDialogFragment.newInstance(..))
doNothing().when(fragment).dismiss()

...
```

(See `ContextMenuFragmentTest`).

Reply:
Thanks I didn't know about `doNothing()`!